### ISSUE TITLE
provide explicit python cache folder

### ISSUE THREAD

**Issue by Andrei-Pozolotin** [kivy/kivy-ios#598]
note: click `▶ Details` sections below to expand the panel with details

1. **this issue was reported before, but was not really resolved**:
* see #425
* see #540

2. **symptom manifests as follows**:
* xcode project works fine on ios/simulator
* xcode project fails when pushed to physical ios/device with `PermissionError` in `.cache`
* problem occurs when python module loader is trying to extract native library, such as `pillow-*.egg`
```
[WARNING] stderr: PermissionError: [Errno 1] Operation not permitted: '/private/var/mobile/Containers/Data/Application/3BB619EB-C6C6-4F25-AC6E-24756016A241/.cache'
```

example physical device log

<details>

```sh
[INFO   ] Logger: Record log in /private/var/mobile/Containers/Data/Application/3BB619EB-C6C6-4F25-AC6E-24756016A241/Documents/.kivy/logs/kivy_21-03-14_1.txt
[INFO   ] Kivy: v2.1.0.dev0
[INFO   ] Kivy: Installed at "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/__init__.py"
[INFO   ] Python: v3.8.2 (default, Mar 14 2021, 00:32:00) 
[Clang 12.0.0 (clang-1200.0.32.29)]
[INFO   ] Python: Interpreter at "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/a-google-tasker"
[INFO   ] Factory: 186 symbols loaded
[INFO   ] KivyMD: 0.104.2.dev0, git-Unknown, 2021-03-14 (installed at "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivymd/__init__.py")
[INFO   ] Image: Providers: img_imageio, img_tex (img_dds, img_sdl2, img_ffpyplayer, img_pil ignored)
[INFO   ] Text: Provider: sdl2
[INFO   ] Window: Provider: sdl2
[INFO   ] GL: Using the "OpenGL ES 2" graphics system
[INFO   ] GL: Backend used <sdl2>
[INFO   ] GL: OpenGL version <b'OpenGL ES 2.0 Metal - 68.8'>
[INFO   ] GL: OpenGL vendor <b'Apple Inc.'>
[INFO   ] GL: OpenGL renderer <b'Apple A9 GPU'>
[INFO   ] GL: OpenGL parsed version: 2, 0
[INFO   ] GL: Shading version <b'OpenGL ES GLSL ES 1.00'>
[INFO   ] GL: Texture max size <4096>
[INFO   ] GL: Texture max units <8>
[INFO   ] Window: auto add sdl2 input provider
[INFO   ] Window: virtual keyboard not allowed, single mode, not docked
[INFO   ] GL: NPOT texture support is available
[INFO   ] main: user_data_dir=/private/var/mobile/Containers/Data/Application/3BB619EB-C6C6-4F25-AC6E-24756016A241/Documents/aelita.google.tasker
[WARNING] stderr: Traceback (most recent call last):
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 1210, in get_cache_path
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 3172, in _bypass_ensure_directory
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 3172, in _bypass_ensure_directory
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 3172, in _bypass_ensure_directory
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 3174, in _bypass_ensure_directory
[WARNING] stderr: PermissionError: [Errno 1] Operation not permitted: '/private/var/mobile/Containers/Data/Application/3BB619EB-C6C6-4F25-AC6E-24756016A241/.cache'
[WARNING] stderr: 
[WARNING] stderr: During handling of the above exception, another exception occurred:
[WARNING] stderr: 
[WARNING] stderr: Traceback (most recent call last):
[WARNING] stderr:   File "/Users/user0/Project/KivyWork/a-google-tasker-ios/YourApp/main.py", line 53, in <module>
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/app.py", line 949, in run
[WARNING] stderr:     self._run_prepare()
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/app.py", line 918, in _run_prepare
[WARNING] stderr:     self.load_kv(filename=self.kv_file)
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/app.py", line 691, in load_kv
[WARNING] stderr:     root = Builder.load_file(rfilename)
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/lang/builder.py", line 306, in load_file
[WARNING] stderr:     return self.load_string(data, **kwargs)
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/lang/builder.py", line 373, in load_string
[WARNING] stderr:     parser = Parser(content=string, filename=fn)
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/lang/parser.py", line 402, in __init__
[WARNING] stderr:     self.parse(content)
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/lang/parser.py", line 508, in parse
[WARNING] stderr:     self.execute_directives()
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivy/lang/parser.py", line 472, in execute_directives
[WARNING] stderr:     mod = __import__(package)
[WARNING] stderr:   File "/Users/user0/Project/KivyWork/a-google-tasker-ios/YourApp/aelita_tasker/view/auth_screen.py", line 3, in <module>
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivymd/uix/__init__.py", line 6, in <module>
[WARNING] stderr:     from kivymd.uix.behaviors import SpecificBackgroundColorBehavior
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivymd/uix/behaviors/__init__.py", line 10, in <module>
[WARNING] stderr:     from .backgroundcolor_behavior import (
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivymd/uix/behaviors/backgroundcolor_behavior.py", line 25, in <module>
[WARNING] stderr:     from .elevation import CommonElevationBehavior
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/kivymd/uix/behaviors/elevation.py", line 377, in <module>
[WARNING] stderr:     from PIL import Image, ImageDraw, ImageFilter
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 618, in _load_backward_compatible
[WARNING] stderr:   File "<frozen zipimport>", line 259, in load_module
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/Pillow-6.1.0-py3.8-macosx-10.15-x86_64.egg/PIL/Image.py", line 95, in <module>
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
[WARNING] stderr:   File "<frozen importlib._bootstrap>", line 618, in _load_backward_compatible
[WARNING] stderr:   File "<frozen zipimport>", line 259, in load_module
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/Pillow-6.1.0-py3.8-macosx-10.15-x86_64.egg/PIL/_imaging.py", line 7, in <module>
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/Pillow-6.1.0-py3.8-macosx-10.15-x86_64.egg/PIL/_imaging.py", line 4, in __bootstrap__
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 1144, in resource_filename
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 1734, in get_resource_filename
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 1763, in _extract_resource
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 1212, in get_cache_path
[WARNING] stderr:   File "/private/var/containers/Bundle/Application/313BBCAC-41BB-4C9A-BBC4-74470CF4B44A/a-google-tasker.app/lib/python3.8/site-packages/setuptools-40.9.0-py3.8.egg/pkg_resources/__init__.py", line 1192, in extraction_error
[WARNING] stderr: pkg_resources.ExtractionError: Can't extract file(s) to egg cache
[WARNING] stderr: 
[WARNING] stderr: The following error occurred while trying to extract file(s)
[WARNING] stderr: to the Python egg cache:
[WARNING] stderr: 
[WARNING] stderr:   [Errno 1] Operation not permitted: '/private/var/mobile/Containers/Data/Application/3BB619EB-C6C6-4F25-AC6E-24756016A241/.cache'
[WARNING] stderr: 
[WARNING] stderr: The Python egg cache directory is currently set to:
[WARNING] stderr: 
[WARNING] stderr:   /private/var/mobile/Containers/Data/Application/3BB619EB-C6C6-4F25-AC6E-24756016A241/.cache/Python-Eggs
[WARNING] stderr: 
[WARNING] stderr: Perhaps your account does not have write access to this directory?
[WARNING] stderr: You can change the cache directory by setting the PYTHON_EGG_CACHE
[WARNING] stderr: environment variable to point to an accessible directory.
[WARNING] stderr: 
```

</details>

3. **the problem arises due to the following**:
* `kivy-os` python module loader [main.m](https://github.com/kivy/kivy-ios/blob/master/kivy_ios/tools/templates/%7B%7B%20cookiecutter.project_name%20%7D%7D-ios/main.m) uses `setuptools` [pkg_resources](https://github.com/pypa/setuptools/blob/dd522e601426ad35644022278ca2a09a7cc542ad/pkg_resources/__init__.py)
* `setuptools`  in turn uses `apidirs`  [module](https://github.com/ActiveState/appdirs/blob/master/appdirs.py)
* `apidirs` can only detect and provide
```
        Mac OS X:   ~/Library/Caches/<AppName>
        Unix:       ~/.cache/<AppName> (XDG default)
        Win XP:     C:\Documents and Settings\<username>\Local Settings\Application Data\<AppAuthor>\<AppName>\Cache
        Vista:      C:\Users\<username>\AppData\Local\<AppAuthor>\<AppName>\Cache
```
* for `ios/simulator` module `apidirs` uses `Unix` default value `~/.cache/` (i.e. can not detect `ios`)
* `~/.cache/` folder is expanded on `simulator` to:
```
/Users/$USER/Library/Developer/CoreSimulator/Devices/$DEVICE_GUID/data/Containers/Data/Application/$APP_GUID/.cache
```
* `~/.cache/` folder on simulator works fine, since that is part of `$USER` context
* `~/.cache/` folder is expanded on `ios/device` to:
```
/private/var/mobile/Containers/Data/Application/$APP_GUID/.cache
```
and this `.cache` folder can not be created since `$APP_GUID/` folder  is `read-only` on the device
* expected writable cache location on a physical device should be `$APP_GUID/Library/Caches/$BUNDLE_ID`

4. **a solution could be to provide an explicit python cache folder**

for example:

### `main.m`

```objective-c
    // expose code identity, i.e. "org.kivy.prog"
    NSString *bundle_id_value = [info objectForKey:(NSString*)kCFBundleIdentifierKey];
    NSString *bundle_id_entry = [NSString stringWithFormat: @"KIVY_BUNDLE_ID=%@", bundle_id_value];
    putenv((char *)[bundle_id_entry UTF8String]);
```

<details>

```objective-c
//
//  main.m
//  a-google-tasker
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#include "Python.h"
#include "/Users/user0/Project/KivyWork/dist/include/common/sdl2/SDL_main.h"
#include <dlfcn.h>

void export_context();
void run_main_setup();

int main(int argc, char *argv[]) {
    int ret = 0;

    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];

    // Change the executing path to YourApp
    chdir("YourApp");

    // Setup environment
    export_context();

    NSString * resourcePath = [[NSBundle mainBundle] resourcePath];
    NSString *python_home = [NSString stringWithFormat:@"PYTHONHOME=%@", resourcePath, nil];
    putenv((char *)[python_home UTF8String]);

    NSString *python_path = [NSString stringWithFormat:@"PYTHONPATH=%@:%@/lib/python3.8/:%@/lib/python3.8/site-packages:.", resourcePath, resourcePath, resourcePath, nil];
    putenv((char *)[python_path UTF8String]);

    NSString *tmp_path = [NSString stringWithFormat:@"TMP=%@/tmp", resourcePath, nil];
    putenv((char *)[tmp_path UTF8String]);

    NSLog(@"Initializing python");
    Py_Initialize();

    wchar_t** python_argv = PyMem_RawMalloc(sizeof(wchar_t *) *argc);
    for (int i = 0; i < argc; i++)
        python_argv[i] = Py_DecodeLocale(argv[i], NULL);
    PySys_SetArgv(argc, python_argv);

    // If other modules are using the thread, we need to initialize them before.
    PyEval_InitThreads();

    // Invoke kivy launcher configurator
    run_main_setup();

    // Search and start main.py
    const char * prog = [
        [[NSBundle mainBundle] pathForResource:@"YourApp/main" ofType:@"pyc"] cStringUsingEncoding:
        NSUTF8StringEncoding];
    NSLog(@"Running main.py: %s", prog);
    FILE* fd = fopen(prog, "r");
    if ( fd == NULL ) {
        ret = 1;
        NSLog(@"Unable to open main.py, abort.");
    } else {
        ret = PyRun_SimpleFileEx(fd, prog, 1);
        if (ret != 0)
            NSLog(@"Application quit abnormally!");
    }

    Py_Finalize();
    NSLog(@"Leaving");

    [pool release];

    // Look like the app still runs even when we left here.
    exit(ret);
    return ret;
}

// This method reads the available orientations from the Info.plist file and
// shares them via an environment variable. Kivy will automatically set the
// orientation according to this environment value, if it exists. To restrict
// the allowed orientation, please see the comments inside.
void export_context() {
    
    NSDictionary *info = [[NSBundle mainBundle] infoDictionary];
    
    NSArray *orientations = [info objectForKey:@"UISupportedInterfaceOrientations"];

    // Orientation restrictions
    // ========================
    // Comment or uncomment blocks 1-3 in order the limit orientation support

    // 1. Landscape only
    // NSString *result = [[NSString alloc] initWithString:@"KIVY_ORIENTATION=LandscapeLeft LandscapeRight"];

    // 2. Portrait only
    // NSString *result = [[NSString alloc] initWithString:@"KIVY_ORIENTATION=Portrait PortraitUpsideDown"];

    // 3. All orientations
    NSString *result = [[NSString alloc] initWithString:@"KIVY_ORIENTATION="];
    for (int i = 0; i < [orientations count]; i++) {
        NSString *item = [orientations objectAtIndex:i];
        item = [item substringFromIndex:22];
        if (i > 0)
            result = [result stringByAppendingString:@" "];
        result = [result stringByAppendingString:item];
    }
    // ========================

    putenv((char *)[result UTF8String]);
    NSLog(@"Available orientation: %@", result);
    
    //
    // expose code identity
    //
    
    // i.e. "org.kivy.prog"
    NSString *bundle_id_value = [info objectForKey:(NSString*)kCFBundleIdentifierKey];
    NSString *bundle_id_entry = [NSString stringWithFormat: @"KIVY_BUNDLE_ID=%@", bundle_id_value];
    putenv((char *)[bundle_id_entry UTF8String]);
    
    // i.e. "Kivy Prog"
    NSString *bundle_name_value = [info objectForKey:(NSString*)kCFBundleNameKey];
    NSString *bundle_name_entry = [NSString stringWithFormat: @"KIVY_BUNDLE_NAME=%@", bundle_name_value];
    putenv((char *)[bundle_name_entry UTF8String]);
    
    // Special environment to prefer .pyo, and don't write bytecode if .py are found
    // because the process will not have a write attribute on the device.
    putenv("PYTHONOPTIMIZE=2");
    putenv("PYTHONDONTWRITEBYTECODE=1");
    putenv("PYTHONNOUSERSITE=1");
    putenv("PYTHONPATH=.");
    putenv("PYTHONUNBUFFERED=1");
    putenv("LC_CTYPE=UTF-8");
    // putenv("PYTHONVERBOSE=1");
    // putenv("PYOBJUS_DEBUG=1");
    
    #ifndef DEBUG
    putenv("KIVY_NO_CONSOLELOG=1");
    #endif
    
}

void run_main_setup() {
    NSBundle *mainBundle = [NSBundle mainBundle];
    NSString *filePath = [mainBundle pathForResource:@"main_setup" ofType:@"py"];
    NSLog(@"Using setup module: %@", filePath);
    NSString *fileText = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:NULL];
    const char *command = [fileText UTF8String];
    PyRun_SimpleString(command);
}
```

</details>

### `main_setup.py`

```python
    # provide explicit python cache folder
    HOME = environ.get("HOME", "/invalid-home-dir")
    BUNDLE = environ.get("KIVY_BUNDLE_ID", "invalid-bundle-id")
    environ["PYTHON_EGG_CACHE"] = f"{HOME}/Library/Caches/{BUNDLE}"
```

<details>

```python
#
# kivy launcher configurator
#

import sys, imp, types
from os import environ
from os.path import exists, join


def report_process_env_var() -> None:
    print("==============")
    for key, value in sorted(environ.items()):
        print(f"   {key} : {value}")
    print("==============")


def has_device() -> bool:
    "detect if running on physical iphone/ipad device"
    return environ.get("HOME", "").startswith("/private/var")


def has_simulator() -> bool:
    "detect if running on virtual device managed by xcode"
    return environ.get("SIMULATOR_DEVICE_NAME") is not None


def setup_python_cache_dir() -> None:
    HOME = environ.get("HOME", "/invalid-home-dir")
    BUNDLE = environ.get("KIVY_BUNDLE_ID", "invalid-bundle-id")
    if has_device() or has_simulator():
        environ["PYTHON_EGG_CACHE"] = f"{HOME}/Library/Caches/{BUNDLE}"
    else:
        pass  # use setuputils/apidirs discovery for Mac


# kivy environment
environ["KIVY_BUILD"] = "ios"
environ["KIVY_NO_CONFIG"] = "1"
# environ["KIVY_NO_FILELOG"] = "1"
environ["KIVY_WINDOW"] = "sdl2"
environ["KIVY_IMAGE"] = "imageio,tex,gif"
environ["KIVY_AUDIO"] = "sdl2"
environ["KIVY_GL_BACKEND"] = "sdl2"

# "True" disables fullscreen and then statusbar is shown
environ["IOS_IS_WINDOWED"] = "False"

try:
    # python 3
    import _imp
    EXTS = _imp.extension_suffixes()
    sys.modules['subprocess'] = types.ModuleType(name='subprocess')
    sys.modules['subprocess'].PIPE = None
    sys.modules['subprocess'].STDOUT = None
    sys.modules['subprocess'].DEVNULL = None
    sys.modules['subprocess'].CalledProcessError = Exception
    sys.modules['subprocess'].check_output = None
except ImportError:
    EXTS = ['.so']

# Fake redirection to supress console output
if environ.get('KIVY_NO_CONSOLE', '0') == '1':

    class stdnone(object):

        def write(self, *args, **kw): pass

        def flush(self, *args, **kw): pass

    sys.stdout = stdnone()
    sys.stderr = stdnone()


# Custom builtin importer for precompiled modules
class CustomBuiltinImporter(object):

    def find_module(self, fullname, mpath=None):
        # print(f'find_module() fullname={fullname} mpath={mpath}')
        if '.' not in fullname:
            return
        if not mpath:
            return
        part = fullname.rsplit('.')[-1]
        for ext in EXTS:
           fn = join(list(mpath)[0], '{}{}'.format(part, ext))
           # print('find_module() {}'.format(fn))
           if exists(fn):
               return self
        return

    def load_module(self, fullname):
        f = fullname.replace('.', '_')
        mod = sys.modules.get(f)
        if mod is None:
            # print('LOAD DYNAMIC', f, sys.modules.keys())
            try:
                mod = imp.load_dynamic(f, f)
            except ImportError:
                # import traceback; traceback.print_exc();
                # print('LOAD DYNAMIC FALLBACK', fullname)
                mod = imp.load_dynamic(fullname, fullname)
            sys.modules[fullname] = mod
            return mod
        return mod

#
#
#


sys.meta_path.insert(0, CustomBuiltinImporter())

setup_python_cache_dir()
report_process_env_var()

```

</details>

**Comment by Afani-Ctrl** [comment]
Does this mean that with Pillow 8.2.0 we will be able to successfully run KivyMD projects on XCODE ios simulators?

**Comment by rajathithan** [comment]
When i try to add the below code in main.m , it gives an error - "use of undeclared identifier info"
'''
// expose code identity, i.e. "org.kivy.prog"
    NSString *bundle_id_value = [info objectForKey:(NSString*)kCFBundleIdentifierKey];
    NSString *bundle_id_entry = [NSString stringWithFormat: @"KIVY_BUNDLE_ID=%@", bundle_id_value];
    putenv((char *)[bundle_id_entry UTF8String]);
'''

**Comment by misl6** [comment]
Should be fixed via #606
